Introduce Expression#canPushQueryToSource and Expression#canPushSortToSource to decide if expression can be executed in lucene

iverase opened the pull request:
Currently this logic is in LocalPhysicalPlanOptimizer using if statements with instanceof checks. This can be achieved in a much cleaner way by moving this responsibility to the Expression class.

iverase committed:
Introduce Expression#canPushToSource to decide if expression can be executed in lucene

elasticsearchmachine commented:
Pinging @elastic/es-analytical-engine (Team:Analytics)

nik9000 commented on x-pack/plugin/esql-core/src/main/java/org/elasticsearch/xpack/esql/core/expression/Expression.java:
```diff
@@ -11,13 +11,17 @@
 import org.elasticsearch.xpack.esql.core.capabilities.Resolvable;
 import org.elasticsearch.xpack.esql.core.capabilities.Resolvables;
+import org.elasticsearch.xpack.esql.core.expression.predicate.operator.comparison.Equals;
+import org.elasticsearch.xpack.esql.core.expression.predicate.operator.comparison.NotEquals;
```
We're kind of in the process of moving everything from `esql-core` to `esql`. If you need some kind of hacky temporary solution you could use reflection. with a big comment that we'll zap it "real soon now"

not-napoleon replied:
I'm very close to deleting the `core` versions of the binary comparison operators; we just need to finish migrating the `In` mapper, and then nothing in ESQL will actually reference them.  Unfortunately, we're a lot farther from being able to pull `Expression` out of core, although that is ultimately a goal.

nik9000 commented on x-pack/plugin/esql-core/src/main/java/org/elasticsearch/xpack/esql/core/expression/predicate/logical/BinaryLogic.java:
```diff
@@ -62,4 +64,9 @@ public Nullability nullable() {
     }
+
+    @Override
+    public boolean canPushToSource(Predicate<FieldAttribute> hasIdenticalDelegate) {
```
Question: should this be `final`? Will subclasses ever override this?

not-napoleon commented on x-pack/plugin/esql-core/src/main/java/org/elasticsearch/xpack/esql/core/expression/Expression.java:
```diff
@@ -207,4 +219,38 @@ public String toString() {
+                // no range or regex queries supported with metadata fields
+                || operation instanceof Equals
+                || operation instanceof NotEquals
+                || operation instanceof WildcardLike;
```
I wonder if instead of checking the instance type here, we should override this on the relevant operators?  That would also work around the fact that we can't see the right classes from here.

iverase replied:
I just did that :)

iverase committed:
iter

iverase committed:
iter

nik9000 commented on x-pack/plugin/esql-core/src/main/java/org/elasticsearch/xpack/esql/core/expression/predicate/operator/comparison/BinaryComparison.java:
```diff
@@ -39,6 +41,11 @@ public DataType dataType() {
     }
 
+    @Override
+    public boolean canPushToSource(Predicate<FieldAttribute> hasIdenticalDelegate) {
```
Same question as above - should be `final`?

nik9000 reviewed (commented):
I love it.

iverase committed:
separate pushdowns from queries and sort

iverase committed:
iter

iverase commented:
I terate again and decided to add two methods, one for push down queries and another for pushing down sorts as they might ( and should) diverge at some point.  I added the provided feedback and removed any instanceof by implementing the method it in all the special cases.

nik9000 reviewed (approved):
I love it.

I think @astefan should have a look too.

not-napoleon commented on x-pack/plugin/esql-core/src/main/java/org/elasticsearch/xpack/esql/core/expression/Expression.java:
```diff
@@ -207,4 +223,34 @@ public String toString() {
+    }
+
+    public interface FieldInfo {
+        boolean hasIdenticalDelegate(FieldAttribute attr);
```
This should probably have some javadoc for what this predicate checks.  Just copying the bit from `PlannerUtils` that's getting deleted would be enough, I think.

iverase replied:
good idea

iverase committed:
Merge branch 'main' into expressionPush

iverase committed:
Add javadocs to FieldInfo

ChrisHegarty reviewed (approved):
LGTM

iverase commented:
got pushed back for this change so closing
